Userland/Libraries/LibCrypto/Hash/SHA1.cpp

```diff
@@ -67,6 +77,86 @@ inline void SHA1::transform(u8 const* data)
     secure_zero(blocks, 16 * sizeof(u32));
 }
 
+#if (ARCH(I386) || ARCH(X86_64))
+// Note: The SHA extension was introduced with
+//       Intel Goldmont (SSE4.2), Ice Lake (AVX512), Rocket Lake (AVX512), and AMD Zen (AVX2)
+//       So it's safe to assume that if we have SHA we have at least SSE4.2
+//      ~https://en.wikipedia.org/wiki/Intel_SHA_extensions
+__attribute__((target("sha,sse4.2"))) static void transform_impl_sha1(u32 (&state)[5], u8 const (&data)[64])
+{
+    using AK::SIMD::i32x4;
+    i32x4 abcd[2] {};
+    i32x4 msgs[4] {};
+
+    abcd[0] = AK::SIMD::load_unaligned<i32x4>(&state[0]);
+    abcd[0] = AK::SIMD::item_reverse(abcd[0]);
+    i32x4 e { 0, 0, 0, static_cast<i32>(state[4]) };
+
+    auto group = [&]<int i_group>() {
+        // FIXME: Test if unrolling the loop is worth it
+        //          GCC: #pragma GCC unroll(5)
+        //        Clang: #pragma unroll
+        for (size_t i_pack = 0; i_pack != 5; ++i_pack) {
+            size_t i_msg = i_group * 5 + i_pack;
+            if (i_msg < 4) {
+                msgs[i_msg] = AK::SIMD::load_unaligned<i32x4>(&data[i_msg * 16]);
+                msgs[i_msg] = AK::SIMD::byte_reverse(msgs[i_msg]);
+            } else {
+                msgs[i_msg % 4] = __builtin_ia32_sha1msg1(msgs[i_msg % 4], msgs[(i_msg + 1) % 4]);
+                msgs[i_msg % 4] ^= msgs[(i_msg + 2) % 4];
+                msgs[i_msg % 4] = __builtin_ia32_sha1msg2(msgs[i_msg % 4], msgs[(i_msg + 3) % 4]);
+            }
+            if (i_msg == 0) {
+                e += msgs[0];
+            } else {
+                e = __builtin_ia32_sha1nexte(abcd[(i_msg + 1) % 2], msgs[(i_msg + 0) % 4]);
+            }
+            abcd[(i_msg + 1) % 2] = __builtin_ia32_sha1rnds4(abcd[(i_msg + 0) % 2], e, i_group);
+        }
+    };
+
+    auto old_abcd = abcd[0];
+    auto old_e = e;
+    group.operator()<0>();
+    group.operator()<1>();
+    group.operator()<2>();
+    group.operator()<3>();
+    e = __builtin_ia32_sha1nexte(abcd[1], i32x4 {});
+    abcd[0] += old_abcd;
+    e += old_e;
+
+    abcd[0] = AK::SIMD::item_reverse(abcd[0]);
+    AK::SIMD::store_unaligned(&state[0], abcd[0]);
+    state[4] = static_cast<u32>(e[3]);
+}
+
+// FIXME: We need a custom resolver as Clang and GCC either refuse or silently ignore the `sha` target
+//        for function multiversioning
+[[gnu::ifunc("resolve_transform_impl")]] static void transform_impl(u32 (&state)[5], u8 const (&data)[64]);
+namespace {
+extern "C" [[gnu::used]] decltype(&transform_impl) resolve_transform_impl()
+{
+    // FIXME: Use __builtin_cpu_supports("sha") when compilers support it
+    constexpr u32 cpuid_sha_ebx = 1 << 29;
+    u32 eax, ebx, ecx, edx;
+    __cpuid_count(7, 0, eax, ebx, ecx, edx);
```

Comment:
We can hide `__cpuid_count` in some `AK::cpu_supports(Capability::SHA)` which would also cache the result. (This will be similar to how `__builtin_cpu_supports` works)

Reply:
Well copilot said the same, but sadly we dont have that API yet

Reply:
It's never late to introduce a new helper!

Reply:
The problem is with clang only, GCC 11.1 supports both the function attribute and the query.

Reply:
Nope, we later figured on Discord that neither Clang nor GCC did the correct thing. Compare assembly for `use_f` and `use_transform_impl` [here](https://godbolt.org/z/GxKhzdPqr).

Reply:
Sorry, I was wrong. GCC does support the `__builtin_cpu_supports("sha")` test since 11.1 but clang does not, not even the latest one version 18. https://godbolt.org/z/cP7WGYqGq Both compilers do not support the automatic dispatch thing (Function Multiversioning).

Reply:
Maybe add test whether the `CPUID` has at least `7` leaves before reading from the `7th` leaf. This is done by setting the `eax` register to `zero` and invoking the `cpuid` instruction, or, from the `C` language, by calling the `__get_cpuid_max(0, NULL)` function provided by the `<cpuid.h>` header. https://godbolt.org/z/7ovcjcM5T

Reply:
> Maybe add test whether the `CPUID` has at least `7` leaves before reading from the `7th` leaf. This is done by setting the `eax` register to `zero` and invoking the `cpuid` instruction, or, from the `C` language, by calling the `__get_cpuid_max(0, NULL)` function provided by the `<cpuid.h>` header. https://godbolt.org/z/v6q89rqhn

For now I'll elide that, as `memset` also avoids that, and I am pretty sure that cpus without the Extentions leaf are missing other features we rely on